### ISSUE TITLE
Python dependency gets wrong category name when pip is only used in dev dependencies.

### ISSUE THREAD

**Issue by ghost** [conda-incubator/conda-lock#275]
Hi,

I have an environment file `environment.yml`:
```python
category: main
dependencies:
  - python=3.10.6
```
And a dev environment `dev-environment.yml`:
```
category: dev
dependencies:
- pip:
   - numpy
```
When I create the lock file (`conda-lock lock -f environment.yml -f dev-environment.yml --check-input-hash --mamba`) the python dependency will now have `category: dev`. This breaks `conda-lock install --no-dev` since python will not be installed.

Probably something in the pip code overwrites the category of python.

**Comment by maresb** [comment]
Hi, I'm having trouble reproducing this. What's your conda-lock version?

BTW, a few minor suggestions: 

1. You may have default channels configured in your `~/.condarc`, but to ensure portability, it's best to be explicit and include
    ```yaml
    channels:
    - conda-forge
    ```
    (we should probably warn about this, since I'm getting a nasty-looking and confusing traceback)

2. Typically one doesn't pin the patch version for something which adheres reasonably well to semver like Python. (In other words, pin `python=3.10` and let conda-lock select the latest patch version.)

**Comment by mariusvniekerk** [comment]
You can also try adding in python as a dep in the dev-environment, that will probably make things behave correctly

**Comment by maresb** [comment]
Last time I tried adding python as a `dev` dep it led to Python not being installed in `main`. That's why we need to turn `category: str` into `categories: list[str]`.

**Comment by mariusvniekerk** [comment]
I would recommend not including `conda-lock` in your lockfile itself.  Thats a rather weird antipattern.

**Comment by maresb** [comment]
> I would recommend not including conda-lock in your lockfile itself. Thats a rather weird antipattern.

Yes, but it's normal to have `conda-lock` in `dev-environment.yml` right? (Especially if one uses `micromamba` to create the environment from the lockfile.)

**Comment by mariusvniekerk** [comment]
I tend to just install conda-lock with pipx or condax and have it live entirely in its own environment.  It's a cli tool so no need to import it or mix it with other stuff

**Comment by voodoo11** [comment]
I can confirm issue exists. It is not only a `python` problem but also applies to other dependencies related to pip packages. Having `conda-lock` in env seems unrelated.

**Comment by voodoo11** [comment]
> Last time I tried adding something as a `dev` and `main` dep it led to it not being installed in `main`. That's why we need to turn `category: str` into `categories: list[str]`.

Basically categories are reliable only if whole dependency trees of categories are disjoint. First, no dependency, direct or indirect, can have two categories, so it is not possible to select an arbitrary subset of categories when installing the environment. Second, a category can be overwritten in an non-deterministic manner, which makes it impossible to create hierarchy of categories.

**Comment by maresb** [comment]
Regarding `categories:`, I proposed a change to the lockfile format in https://github.com/mamba-org/mamba/issues/1209#issuecomment-1236086736. The response was positive, but there doesn't seem to be any movement yet towards an implementation.

**Comment by maresb** [comment]
Regarding `categories` I just had a rough idea how we might be able to move the concept forward from purely the conda-lock side, posted as #278. I'm not sure if it's viable, but if it works it'd be fairly simple to implement.

**Comment by Ben-Habermeyer** [comment]
I seem to be experiencing this issue, specifically when locking pip requirements defined in a `pyproject.toml` file where dev dependencies are installed in the `[tool.poetry.dev-dependencies]` section.

What I observe is that transitive dependencies managed with pip are installed in the the `dev` category only, such that when installing only main dependencies into an environment, `ModuleNotFound` error is raised by the code. 

Example, requirement X and dev requirement Y both depend on Z, but Z gets installed in dev only.

I am only observing this issue when moving to the new lockfile format (conda-lock >=2)

Is the only solution to assembly 2 separate environments for "dev" and "no dev" where "dev" contains all the dependencies of the latter, plus dev requirements (and therefore not use categories feature)?

**Comment by maresb** [comment]
Hi @Ben-Habermeyer, in practice I've usually managed to work around the issue by ensuring that main dependencies are never repeated as dev dependencies and then occasionally adding any missing transitive dependencies as explicit main dependencies as needed. It's pretty ugly and fragile but usually gets the job done.

Since you're working with a Poetry-based project, I'd highly recommend looking into [pixi](https://github.com/prefix-dev/pixi). Pixi is basically Poetry for conda. It does pypi dependencies a lot more robustly and is being developed actively by a team that I really admire. There are a few differences: it's project-based rather than global-environment-based. Even though you're working with a pyproject.toml I'd recommend using pixi.toml for better conceptual separation even though they're equivalent.

**Comment by Ben-Habermeyer** [comment]
> Hi @Ben-Habermeyer, in practice I've usually managed to work around the issue by ensuring that main dependencies are never repeated as dev dependencies and then occasionally adding any missing transitive dependencies as explicit main dependencies as needed. It's pretty ugly and fragile but usually gets the job done.
> 
> Since you're working with a Poetry-based project, I'd highly recommend looking into [pixi](https://github.com/prefix-dev/pixi). Pixi is basically Poetry for conda. It does pypi dependencies a lot more robustly and is being developed actively by a team that I really admire. There are a few differences: it's project-based rather than global-environment-based. Even though you're working with a pyproject.toml I'd recommend using pixi.toml for better conceptual separation even though they're equivalent.

Thanks @maresb, yes I have been looking at #278, #434 and https://github.com/mamba-org/mamba/issues/1209 as well - seems to be exactly what I am experiencing.

Yea essentially what I have is a bioinformatic python package where dependencies are locked using `pyproject.toml`. Conda-only, non-python packages such as `bwa` are listed under `[tool.conda-lock.dependencies]`, python dependencies under `[tool.poetry.dependencies]` and dev dependencies under `[tool.poetry.dev-dependencies]`. Those which can only be installed via pip (such as our internal pypi) are specified with {"source" = "pypi"}

Previously (conda lock v1) `conda-lock` would generate a multiplatform lock file with `main` and `dev` dependencies separated - dev would be installed for local development (things like linters, pytest etc.) and `main` would be installed on the production image.

But as myself and others have explained, there seems to be a transitive dependency issue now with categories where the transitive dependencies (seems limited to pip dependencies?) are only listed in the `dev` channel.

I have tried your suggestion of adding explicit transitive dependencies with some success. For non-pyproject-toml based requirements, I have also used https://github.com/conda-incubator/conda-env-builder to define environment yaml files which inherit form one another (in this example, `dev` would inherit from `main`). 

I will also take a look at pixi as you suggested. For the time being do you suggest following #273 as the root issue when a fix may be addressed in `conda-lock`? Thanks again.
